mpas_analysis/shared/climatology/climatology.py

```diff
@@ -480,11 +494,178 @@ def compute_climatology(ds, monthValues, calendar=None):
+
+    if printProgress:
+        print '   Computing and caching climatologies covering {}-year ' \
+              'spans...'.format(yearsPerCacheFile)
+
+    cacheIndices = numpy.zeros(ds.dims['Time'], int)
+    yearsInDs = ds.year.values
+
+    # figure out which files to load and which years go in each file
+    for firstYear in range(startYearClimo, endYearClimo+1, yearsPerCacheFile):
+        years = range(firstYear, numpy.minimum(endYearClimo+1,
+                                               firstYear+yearsPerCacheFile))
+        if len(years) == 0:
+            continue
+
+        if yearsPerCacheFile == 1:
+            yearString = '{:04d}'.format(years[0])
+            outputFileClimo = '{}_year{}.nc'.format(cachePrefix, yearString)
+        else:
+            yearString = '{:04d}-{:04d}'.format(years[0], years[-1])
+            outputFileClimo = '{}_years{}.nc'.format(cachePrefix, yearString)
+
+        done = False
+        if os.path.exists(outputFileClimo):
+            # already cached
+            dsCached = xr.open_dataset(outputFileClimo)
+            if dsCached.attrs['totalMonths'] == 12*len(years):
+                # also complete, so we can move on
+                done = True
+
+        cacheIndex = len(cacheInfo)
+        for year in years:
+            cacheIndices[yearsInDs == year] = cacheIndex
+
+        if numpy.count_nonzero(cacheIndices == cacheIndex) == 0:
+            continue
+
+        cacheInfo.append((outputFileClimo, done, years))
+
+    ds = ds.copy()
+    ds.coords['cacheIndices'] = ('Time', cacheIndices)
+
+    # compute and store each cache file
+    for cacheIndex, info in enumerate(cacheInfo):
+        outputFileClimo, done, years = info
+        if done:
+            continue
+        dsYear = ds.where(ds.cacheIndices == cacheIndex, drop=True)
+
+        if printProgress:
+            if yearsPerCacheFile == 1:
+                yearString = '{:04d}'.format(years[0])
+            else:
+                yearString = '{:04d}-{:04d}'.format(years[0], years[-1])
+            print '     {}'.format(yearString)
+
+        totalDays = dsYear.daysInMonth.sum(dim='Time').values
+
+        monthCount = dsYear.dims['Time']
+
+        climatology = compute_climatology(dsYear,  monthValues, calendar,
+                                          maskVaries=False)
+
+        climatology.attrs['totalDays'] = totalDays
+        climatology.attrs['totalMonths'] = monthCount
+
+        climatology.to_netcdf(outputFileClimo)
+
+    # compute the aggregate climatology
+    yearString = '{:04d}-{:04d}'.format(startYearClimo, endYearClimo)
+    outputFileClimo = '{}_years{}.nc'.format(cachePrefix, yearString)
+
+    done = False
+    if os.path.exists(outputFileClimo):
+        climatology = xr.open_dataset(outputFileClimo)
+        if climatology.attrs['totalMonths'] == (endYearClimo-startYearClimo+1)*12:
+            # also complete, so we can move on
+            done = True
+
```

Comment:
This is a super, super long function.  Breaking it up into smaller sub-functions will make it easier to understand.